blockchain/stakeversion.go

```diff
@@ -0,0 +1,359 @@
+		"not found")
+)
+
+// calcWantHeight calculates the height of the final block of the previous
+// interval.  The adjusted height accounts for the fact the starting validation
+// height does not necessarily start on an interval and thus the intervals
+// might not be zero-based.
+func calcWantHeight(skip, interval, height int64) int64 {
+	intervalOffset := skip % interval
+	adjustedHeight := height - intervalOffset - 1
+	return (adjustedHeight - ((adjustedHeight + 1) % interval)) +
+		intervalOffset
+}
+
+// findStakeVersionPriorNode walks the chain backwards from prevNode until it
+// finds the appropriate interval block to return to the caller.
+//
+// This function MUST be called with the chain state lock held (for writes).
+func (b *BlockChain) findStakeVersionPriorNode(prevNode *blockNode) (*blockNode, error) {
+	// Check to see if the blockchain is high enough to begin accounting
+	// stake versions.
+	nextHeight := prevNode.height + 1
+	if nextHeight < b.chainParams.StakeValidationHeight+
+		b.chainParams.StakeVersionInterval {
+		return nil, errStakeVersionHeight
+	}
+
+	wantHeight := calcWantHeight(b.chainParams.StakeValidationHeight,
+		b.chainParams.StakeVersionInterval, nextHeight)
+
+	// Walk backwards until we find an interval block and make sure we
+	// don't blow through the minimum height.
+	iterNode := prevNode
+	for iterNode.height > wantHeight {
+		var err error
+		iterNode, err = b.getPrevNodeFromNode(iterNode)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return iterNode, nil
+}
+
+// isVoterMajorityVersion determines if minVer requirement is met based on
+// prevNode.  The function always uses the voter versions of the prior window.
+// For example, if StakeVersionInterval = 11 and StakeValidationHeight = 13 the
+// windows start at 13 + (11 * 2) 25 and are as follows: 24-34, 35-45, 46-56 ...
+// If height comes in at 35 we use the 24-34 window, up to height 45.
+// If height comes in at 46 we use the 35-45 window, up to height 56 etc.
+//
+// This function MUST be called with the chain state lock held (for writes).
+func (b *BlockChain) isVoterMajorityVersion(minVer uint32, prevNode *blockNode) bool {
+	// Walk blockchain backwards to calculate version.
+	node, err := b.findStakeVersionPriorNode(prevNode)
+	if err != nil {
+		if err == errStakeVersionHeight {
+			return 0 >= minVer
+		}
+		return false
+	}
+
+	// Calculate version.
+	totalVotesFound := int32(0)
+	versionCount := int32(0)
+	iterNode := node
+	for i := int64(0); i < b.chainParams.StakeVersionInterval && iterNode != nil; i++ {
+		totalVotesFound += int32(len(iterNode.voterVersions))
+		for _, version := range iterNode.voterVersions {
+			if version >= minVer {
+				versionCount += 1
+			}
+		}
+
+		var err error
+		iterNode, err = b.getPrevNodeFromNode(iterNode)
+		if err != nil {
+			return false
+		}
+	}
+
+	// Ensure we have enoguh votes.
+	if int64(totalVotesFound) < b.chainParams.StakeVersionInterval*
+		int64(b.chainParams.TicketsPerBlock-2) {
```

Comment:
This can't be hard coded.  How do you know it's `-2`?  What if the `TicketsPerBlock` parameter is 6?